### ISSUE TITLE
Segfault: warp-ctc pytorch bindings

### ISSUE THREAD

**Issue by jerinphilip** [SeanNaren/deepspeech.pytorch#90]
Hello, I'm trying to use CTC for an OCR which uses a similar architecture, following the examples at https://github.com/SeanNaren/warp-ctc/blob/pytorch_bindings/pytorch_binding/README.md.

I was initially having a problem with the types (IntTensor/FloatTensor), which I believe I fixed with some trial and error, but right now I'm stuck at a segfault.

Can you tell me what probs_sizes, and label sizes are? I don't understand [this](https://github.com/SeanNaren/deepspeech.pytorch/blob/master/train.py#L230) in deepspeech.pytorch:

```
 sizes = Variable(input_percentages.mul_(int(seq_length)).int(), requires_grad=False)
```

**Comment by SeanNaren** [comment]
That line represents the length of each output in the output tensor, so that the loss calculation can ignore anything outside that length (due to the padding).

The reason it's done like this is because after the convolution layers, the sequence is compressed (smaller than the input). By taking the output lengths, then multiplying it by percentages (that represent the size of each sample in the batch relative to the largest size) we get the size of each sample in the output batch

**Comment by jerinphilip** [comment]
I'm passing a Wx32 image to the model, where W is the temporal axis and the network outputs probabilities for each class. So my output is  a probability vector of 1x output_class_size for a 1x32 input.  All concatenated should get me a W x output_class_size probability matrix. Assuming a batch size of 1, I'm passing a 
```
# Input Dims: torch.Size([1, 208, 32])
# Output Dims: torch.Size([1, 208, 187])

prob_sizes = [ output_class_size for i in range(W)] 
label_sizes = [ len(target_string) ]
```
converted into torch Tensors.
What could be leading to the segfault, running it through valgrind didn't get me anything. Does the ordering of in the tensor affect computations?

I've found a TxBxH representation in your codes, which I believe stands for seq_length, batch_size, and input_size. Need I reshape the tensor someway to get the code to work?

**Comment by jerinphilip** [comment]
Any reason to get a nan loss?

**Comment by SeanNaren** [comment]
Sorry for the late response, but yes the input to the CTC loss should be `seq_length, batch_size, input_size`

**Comment by ahmedmazari-dhatim** [comment]
Hi @SeanNaren,

1) What is the loss function used here ? is it hing loss, cross-entropy or negative log likelihood 
2) How can l get the probabilities outputted ? since rnn output logits 

Thank you

**Comment by ryanleary** [comment]
1. It is CTC loss. See the Graves 2006 paper: http://www.cs.toronto.edu/~graves/icml_2006.pdf
2. The matrix output from the NN contains the log-probabilities of each character class at each timestamp for each sample in the batch. Are you asking for the probability of the emitted sequence?

**Comment by ryanleary** [comment]
Closing for now, will reopen if further questions.

**Comment by LiangHao92** [comment]
@ryanleary hello， I wonder where is the folder named "pytorch_binding". In the warp-ctc folder I download, there is only folder named "torch_binding" and "ttensorflow_binding". where is the torch_binding folder I can binding pytorch?

**Comment by ryanleary** [comment]
You need to use @SeanNaren 's fork: https://github.com/SeanNaren/warp-ctc

**Comment by gentaiscool** [comment]
Is it possible to get the probability of the emitted sequences?
